proposals/1113_ldap-ad-web-ui/README.md

```diff
@@ -0,0 +1,136 @@
+* Provide generalized error messaging for failed logins that do not expose sensitive information (i.e. we should say "invalid username or password"
+  but do not expose whether it's the username or password that's incorrect)
+* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)
+* Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)
+* Prevent LDAP injection attacks
+* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
+* Screens are accessible and friendly to screen readers
+* Screens are friendly to password managers
+
+#### Non-goals
+* A rich client (ie the use of javascript)
+* Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.
+* Branding & customization beyond the information listed in the goals used to identify the login screen belongs to Pinniped.
+* Supporting SSO integrations
+* Internationalization or localization. The CLI doesn't currently support this either.
+
+### Specification / How it Solves the Use Cases
+
+#### API Changes
+
+The supervisor must accept requests from other clients, as detailed
+in the (todo) proposal for dynamic client registration.
+When a client other than pinniped-cli makes an authorization endpoint request with `response_type=code` and their
+IDP is an LDAP or Active Directory IDP, the user will be redirected to the new login page.
+When a client other than the Pinniped CLI makes an authorization endpoint request with
+custom Username/Password headers, they should be rejected.
+
+The pinniped cli should default to using the cli-based password flow, but when a tty is unavailable or when
+a CLI flag (`--ldap-browser-login`? `--force-browser-login`?) is specified, it will open a browser to log in
+instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
+may wish to authenticate using the pinniped cli but without access to a terminal.
+
+#### Upgrades
+
+This change is backwards compatible. Users would see no changes unless they decided to register
+a new client or change the pinniped cli flags.
+
+However if they do choose to register a new client they may need to update the following:
+- FederationDomains today may be using private certificate authorities. These are trusted
+  for our use case but a browser will flag them as unsafe. Admins will have to transition to letsencrypt
+  or another public Certificate Authority to prevent making end users click past messages about the certificate
+  being untrusted.
+- The name of the idp custom resource is currently not published to users logging in with Pinniped.
+  We plan on exposing this to indicate to users which idp they are logging in to.
+  Admins may need to update this to something more user-friendly.
```

Comment:
It is part of our custom discovery metadata IIRC.

Reply:
It's always been available but not advertised. Now it will be advertised, which seemed worth calling out.

Reply:
Do we really need to expose the name?  You can only have one IDP configured, so if the IDP name did not matter for the CLI, why does it matter in the UI?

Reply:
Honestly I was wondering the same thing, the document that Jeremy and Anjali came up with had it as one of their goals which is why I put it here as one. 
Perhaps we are expecting admins to have communicated instructions about what IDP they're logging in to when they distribute the admin kubeconfig? CLI users may be primed about how to use Pinniped in a way that users of dashboards might not be.

Reply:
Some things to consider:
* This proposal will result in a minimally branded login page. That branding will almost definitely not match enterprise branding.  While it isn’t much, providing the IdP name offers a small clue the user that they're in the right place and not some place trying to steal their password.
* Multi IdP will eventually arrive, might as well start prepping.
* I imagine the CLI being brought in line to show the IdP name too.

Reply:
> Admins may need to update this to something more user-friendly.

This seems like a good idea. Without it, we could only assume that we should use the name of the IDP CR, which is not allowed to be formatted very well (e.g. no spaces). I could imagine adding this as a new optional field in the OIDC/LDAP/ActiveDirectoryIdentityProvider CRDs. When set, the name could be shown in the new LDAP/AD login UI, and shown by the CLI above the login prompts for LDAP/AD and OIDC password grant. When not set, those places could show the name of the IDP CR.

I could also imagine adding a similar field in the same places to allow the admin to rename the username prompt (e.g. name it "Email Address"). This may be outside the scope of this proposal, but we could leave room for it if we add a new API field.

All of these optionally configured strings could be added to the IDP discovery result to be consumed by `pinniped get kubeconfig` and encoded into the kubeconfig as CLI flags.

Maybe allowing configuring a login banner message could fit into the same place someday too.

Reply:
Resolved, see next sentence about leaving further branding decisions until later.